### PR TITLE
formatting of debug output of traces

### PR DESCRIPTION
**Pull request by kroening**
This does minor tweaks to the debug output of goto_traces; in particular, this now uses format().

### CODE REVIEW

File: src/goto-programs/goto_trace.cpp [hunk 1]
```diff
@@ -149,6 +134,10 @@ std::string trace_value_binary(
       if(!to_integer(expr, i) && i>=0)
         return integer2string(i, 2);
     }
+    else if(type.id()==ID_string)
+    {
+      return integer2string(expr.get(ID_value).get_no(), 2);
```
Comment (peterschrammel) [inline src/goto-programs/goto_trace.cpp]: This doesn't work with `USE_STD_STRING` enabled: https://travis-ci.org/diffblue/cbmc/jobs/378446491#L1421

File: src/goto-programs/goto_trace.cpp [hunk 2]
```diff
@@ -61,61 +62,45 @@ void goto_trace_stept::output(
...
-  else if(pc->is_decl())
-    out << "DECL   ";
-  else if(pc->is_function_call())
-    out << "CALL   ";
-  else
-    out << "(?)    ";
+    out << pc->source_location << '\n';
 
-  out << "\n";
+  out << pc->type << '\n';
```
Comment (smowton) [inline src/goto-programs/goto_trace.cpp:85]: Why is this one changed but the switch statement at the top of this method isn't?

**Review by smowton — COMMENTED** [review]
One question, plus should we be outputting something generic but useful for the other instruction types instead of just printing "SHARED_WRITE" by itself?

File: src/goto-programs/goto_trace.h [hunk 3]
```diff
@@ -104,14 +104,18 @@ class goto_trace_stept
   // for assert
   std::string comment;
 
-  // the object being assigned
+  // The object being assigned.
+  // This will get removed, use full_lhs instead.
   ssa_exprt lhs_object;
```
Comment (tautschnig) [inline src/goto-programs/goto_trace.h]: What is blocking immediate removal?

File: src/goto-programs/goto_trace.cpp (hunk 2 above)
Comment (kroening) [reply src/goto-programs/goto_trace.cpp:85]: Switching over something different!

File: src/goto-programs/goto_trace.cpp [hunk 4]
```diff
@@ -61,61 +62,45 @@ void goto_trace_stept::output(
...
+        out << "  " << pc->source_location << '\n';
 
-      if(comment!="")
-        out << "  " << comment << "\n";
-      out << "  " << from_expr(ns, pc->function, pc->guard) << "\n";
-      out << "\n";
+      if(!comment.empty())
+        out << "  " << comment << '\n';
+
+      out << "  " << format(pc->guard) << '\n';
```
Comment (tautschnig) [inline src/goto-programs/goto_trace.cpp:98]: ~I'm not convinced there is value in making the debug output language-independent when the plain-text user-directed output is and should remain language-specific?~  Clarified by @peterschrammel's comment.

**Comment by peterschrammel** [comment]
I think that we should make a clear distinction between output that is GOTO and output that is language-specific. We are sometimes a bit fooled by the similarity of GOTO and C, but for Java this is no more the case and language-specific output is indeed required by downstream tools. This mostly concerns the structured (JSON, XML) outputs of goto traces. (Non-front-end) debug output should always be GOTO. For all other output functions, we should provide them with a formatter function supplied by the driver program, as @kroening suggests. The driver program can then decide whether to call format() or a language-specific version of it. Hard-coded calls to from_expr/format should be replaced by that.

**Comment by kroening** [comment]
To add, some of the "debug output" methods could likely be replaced by a "generic output" method that then gets given format_expr as the formatter. On the side, we will eventually need a uniform name for a "debug output" method.

File: src/goto-programs/goto_trace.h (hunk 3 above)
Comment (kroening) [reply src/goto-programs/goto_trace.h]: They are still being output to the XML output, and I am unsure whether there are still any users of this. I have been warning about this for ~3 years.

Comment (kroening) [reply src/goto-programs/goto_trace.h]: The motivation for removal is that the full object may be huge (think of a large array of a big struct). Various users have raised the sheer size of the XMLs as a concern.

Comment (tautschnig) [reply src/goto-programs/goto_trace.h]: The introduced comment suggests that there equivalent replacements. If that isn't the case (which is what your comments here suggest), then the comment needs to be fixed.

Comment (kroening) [reply src/goto-programs/goto_trace.h]: They are equivalent in the sense that you can reconstruct the lhs_object value from the full_lhs values; same for lhs_object and lhs.

Comment (tautschnig) [reply src/goto-programs/goto_trace.h]: So ... that should be done?

Comment (kroening) [reply src/goto-programs/goto_trace.h]: Ok, will drop this commit, and will do a separate PR for the removal.

Comment (tautschnig) [reply src/goto-programs/goto_trace.h]: Makes a lot of sense, thanks!

**Review by tautschnig — APPROVED** [review]

**Review by smowton — APPROVED** [review]

**Pull request merged**